File: apps/dashboard/src/lib/assets.ts
import type { AssetClass } from "@kronos/shared-types";

/** Alpaca crypto pairs use slash form (BTC/USD). */
export function isCryptoSymbol(symbol: string): boolean {
  return symbol.includes("/");
}

export function assetClassOf(symbol: string): AssetClass {
  return isCryptoSymbol(symbol) ? "crypto" : "us_equity";
}

/** Normalize free-typed tickers; BTCUSD → BTC/USD. */
export function normalizeSymbol(raw: string): string {
  const s = raw.trim().toUpperCase().replace(/\s+/g, "");
  if (!s) return s;
  if (s.includes("/")) {
    const parts = s.split("/").filter(Boolean);
    if (parts.length === 2) return `${parts[0]}/${parts[1]}`;
    return s;
  }
  const quotes = ["USDT", "USDC", "USD", "BTC", "ETH"] as const;
  for (const quote of quotes) {
    if (s.endsWith(quote) && s.length > quote.length + 1) {
      const base = s.slice(0, -quote.length);
      if (/^[A-Z]{2,6}$/.test(base)) return `${base}/${quote}`;
    }
  }
  return s;
}

export function formatPrice(px: number, symbol?: string): string {
  if (!Number.isFinite(px)) return "—";
  if (symbol && isCryptoSymbol(symbol)) {
    if (px >= 1000) return px.toFixed(2);
    if (px >= 1) return px.toFixed(4);
    return px.toFixed(6);
  }
  return px.toFixed(2);
}

File: apps/trader/app/assets.py
from __future__ import annotations

# Quote currencies Alpaca crypto pairs use — used to normalize BTCUSD → BTC/USD.
_CRYPTO_QUOTES = ("USDT", "USDC", "USD", "BTC", "ETH")


def normalize_symbol(raw: str) -> str:
    """Canonical ticker: uppercased; crypto pairs keep or gain a slash (BTC/USD)."""
    s = raw.strip().upper().replace(" ", "")
    if not s:
        return s
    if "/" in s:
        parts = [p for p in s.split("/") if p]
        if len(parts) == 2:
            return f"{parts[0]}/{parts[1]}"
        return s
    for quote in _CRYPTO_QUOTES:
        if s.endswith(quote) and len(s) > len(quote) + 1:
            base = s[: -len(quote)]
            if base.isalpha() and 2 <= len(base) <= 6:
                return f"{base}/{quote}"
    return s


def is_crypto_symbol(symbol: str) -> bool:
    """Alpaca crypto trading pairs always contain '/' (e.g. BTC/USD)."""
    return "/" in (symbol or "")


def asset_class_of(symbol: str) -> str:
    return "crypto" if is_crypto_symbol(symbol) else "us_equity"


def partition_symbols(symbols: list[str]) -> tuple[list[str], list[str]]:
    """Return (equity_symbols, crypto_symbols) preserving order."""
    equity: list[str] = []
    crypto: list[str] = []
    for s in symbols:
        if is_crypto_symbol(s):
            crypto.append(s)
        else:
            equity.append(s)
    return equity, crypto
